lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1098 @@
+        use ReaderCommand::*;
+        use ReaderState::*;
+
+        match (cmd, &self.state) {
+            (_, &ShuttingDown) => {}
+            (PauseRead, _) => self.state = Paused,
+            (ResumeRead, _) => self.state = Reading,
+            (Shutdown, _) => self.state = ShuttingDown,
+        }
+    }
+
+    /// Blocks on read() and handles the response accordingly.
+    fn do_read(&mut self, buf: &mut [u8; 8192]) {
+        use std::io::ErrorKind::*;
+        use ReaderState::*;
+
+        match self.inner.read(buf) {
+            Ok(0) => {
+                // Peer disconnected or TcpStream::shutdown was called.
+                // Notify the PeerManager then shutdown
+                self.peer_manager.socket_disconnected(&self.descriptor);
+                self.peer_manager.process_events();
+                self.state = ShuttingDown;
+            }
+            Ok(bytes_read) => {
+                match self
+                    .peer_manager
+                    .read_event(&mut self.descriptor, &buf[0..bytes_read])
+                {
+                    Ok(pause_read) => {
+                        if pause_read {
+                            self.state = Paused;
+                        }
+                    }
+                    Err(_) => {
+                        // Rust-Lightning told us to disconnect;
+                        // no need to notify PeerManager in this case
+                        self.state = ShuttingDown;
+                    }
+                }
+
+                self.peer_manager.process_events()
+            }
+            Err(e) => match e.kind() {
+                TimedOut | Interrupted => {
+                    // Acceptable error; retry
+                }
+                _ => {
+                    // For all other errors, notify PeerManager and shut down
+                    self.peer_manager.socket_disconnected(&self.descriptor);
+                    self.peer_manager.process_events();
+                    self.state = ShuttingDown;
+                }
+            },
+        }
+    }
+}
+
+/// An actor that synchronously initiates the write() events requested by the
+/// `PeerManager`.
+struct Writer<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    inner: TcpWriter,
+    peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+    descriptor: SyncSocketDescriptor,
+    writer_cmd_rx: Receiver<WriterCommand>,
+    write_data_rx: Receiver<Vec<u8>>,
+    /// An internal buffer which stores the data that the Writer is
+    /// currently attempting to write.
+    ///
+    /// This buffer is necessary because calls to self.inner.write() may fail or
+    /// may write only part of the data.
+    buf: Option<Vec<u8>>,
```

Comment:
> I think you should be able to do the writer with no separate thread at all, either!

If I'm understanding you correctly, you mean to use `TcpStream`'s `set_nonblocking(true)`, and then do `stream.write()` directly within `SyncSocketDescriptor::send_data()`? That would work for the writer, but since the `Reader` and `Writer` share the same underlying `TcpStream` which is managed by the OS, this means that the reader would no longer be able to block on `read()`, and could only read in a busy loop. From the [`set_nonblocking()` docs](https://doc.rust-lang.org/std/net/struct.TcpStream.html#method.set_nonblocking):

> This will result in `read`, `write`, `recv` and `send` operations becoming nonblocking, i.e., immediately returning from their calls. If the IO operation is successful, `Ok` is returned and no further action is required. If the IO operation could not be completed and needs to be retried, an error with kind [`io::ErrorKind::WouldBlock`](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WouldBlock) is returned.
> 
> On Unix platforms, calling this method corresponds to calling `fcntl FIONBIO`. On Windows calling this method corresponds to calling `ioctlsocket FIONBIO`

Reply:
> rust-lightning should re-call with the remaining data if you only write part

This requires `send_data()` to know immediately whether a call to `write()` wrote all of the bytes, some of the bytes, etc. But since we cannot use non-blocking writes due to the constraints on the `Reader` above, there needs to be a dedicated `Writer` thread that can block on `write()`. Which AFAICT necessitates a layer of indirection (i.e. the `write_data` channel) between `send_data()` and the `Writer`, which allows the `Writer` to receive write requests while it is blocked on `write(), but which unfortunately makes partial `write()`s undetectable to callers of `send_data()`.

Reply:
> I think you should be able to do the writer with no internal buffer at all

I could imagine making the `write_data` channel a zero-capacity channel, so that when the `Writer` `recv()`s data, it immediately turns around and passes the data into `write()`. But since `send_data()` cannot detect partial writes (as explored above), the `Writer` must continue trying to send the remaining data until it succeeds (or shuts down), since the call to `send_data()` reported to Rust-Lightning that the entire write was a success. And in order to do that, the remaining data must be stored in memory somewhere, thus the internal buffer.

I don't see a way around these constraints - any suggestions? The constraints are subtle but have wide-reaching implications regarding the design of this crate.

Reply:
If we set a timeout and `write()` times out, there's no way for the `PeerManager` to know that there is `write_buffer_space_avail()`. After the `Writer` completes a `write()`, it `recv()`s the next `Vec<u8>`, at which point it can notify the `PeerManager`. But without futures or polling, a failed `send_data()` call does not have a mechanism for making that call. Yet another wrinkle of using 100% synchronous code...

Also, timeouts [aren't configured at all](https://edp.fortanix.com/docs/concepts/rust-std/) when compiling to the EDP target.

> The following functions don't configure a timeout at all and revert back to default behavior:
>
> `net::TcpStream::connect_timeout`
> `net::TcpStream::set_read_timeout`
> `net::TcpStream::set_write_timeout`

This would mean that the future `Socket` trait would implementing these functions too. Right now, we are using only the most basic and common POSIX APIs - `connect`, `read`, `write`, `shutdown`. The more advanced functionality we use, the more functions we need to include into the `Socket` trait, which in turn means more burden on users that need to make the `lightning-net` work on their specific platform, possibly highly constrained platform which doesn't have access to as rich of an API set from the OS.

Reply:
For example, [this POSIX socket](https://docs.rs/posix-socket/latest/posix_socket/struct.Socket.html) does not support `connect_timeout`.

Reply:
The `Writer` channels could be replaced with something like `Arc<(Mutex<(Option<Vec<u8>>, bool)>, Condvar)>` (this is my first exposure to condition variables) but I don't see how we can get rid of the `Writer` thread.

Reply:
Refreshed and only just saw your last four messages

> This is kinda nonsense, though - the os can always "deny service to an enclave" but just.....not context-switching to that process? Like, what does TCP timeouts have to do with it...

Yeah, it's a pain... I don't know what Fortanix's exact rationale is, but as the ecosystem currently stands the alternatives are to (1) make `Tokio` run inside SGX by fixing the entire Tokio stack down to `mio` or (2) to patch Rust `std`.

> Two threads per connection is a lot of overhead.

FWIW, for our use case it's not a problem at all, since our users only really need to connect to an LSP. I figure that if someone truly needs performance, they would be using Tokio and therefore `lightning-net-tokio`. This crate can serve people who need LDK to run on uncommon or not-well-supported platforms. And the way to accomplish that IMO is to keep the API to the OS extremely simple.

Reply:
I'll take a stab at replacing all of the `crossbeam` channels with shared state + condition variables, which should also eliminate the `select!`, and therefore the entire dependency. Beyond that, the only thing left to make this run inside SGX is to implement the custom usercalls. So I don't think there's that much more work to do actually.

Reply:
> To rethink this somewhat further, instead of jumping through hoops here, we could also consider running the actual net logic outside of the enclave - it doesn't change the security model to make the syscalls outside of the enclave, we just pass the stream of read/write_buffer_space_avail events and write calls over some IPC. I'd bet we could do all that with less total LoC :/

That's essentially what we will be doing once the usercalls are implemented: making the syscalls outside of the enclave. Maybe take a quick look at [the EDP architecture](https://edp.fortanix.com/docs/concepts/architecture/) (it's short)